tests/ga/test_exthandlers_download_extension.py

```diff
@@ -0,0 +1,203 @@
+        finally:
+            if file is not None:
+                file.close()
+
+    @staticmethod
+    def _create_invalid_zip_file(filename):
+        with open(filename, "w") as file:
+            file.write("An invalid ZIP file\n")
+
+    def _get_extension_package_file(self):
+        return os.path.join(self.agent_dir, os.path.basename(self.pkg.uris[0].uri) + ".zip")
+
+    def _get_extension_command_file(self):
+        return os.path.join(self.extension_dir, DownloadExtensionTestCase._extension_command)
+
+    def _assert_download_and_expand_succeeded(self):
+        self.assertTrue(os.path.exists(self._get_extension_package_file()), "The extension package was not downloaded to the expected location")
+        self.assertTrue(os.path.exists(self._get_extension_command_file()), "The extension package was not expanded to the expected location")
+
+    def test_it_should_download_and_expand_extension_package(self):
+        def download_ext_handler_pkg(_uri, destination):
+            DownloadExtensionTestCase._create_zip_file(destination)
+            return True
+
+        with patch("azurelinuxagent.common.protocol.wire.WireProtocol.download_ext_handler_pkg", side_effect=download_ext_handler_pkg) as mock_download_ext_handler_pkg:
+            self.ext_handler_instance.download()
+
+        # first download attempt should succeed
+        mock_download_ext_handler_pkg.assert_called_once()
+
+        self._assert_download_and_expand_succeeded()
+
+    def test_it_should_use_existing_extension_package_when_already_downloaded(self):
+        DownloadExtensionTestCase._create_zip_file(self._get_extension_package_file())
+
+        with patch("azurelinuxagent.common.protocol.wire.WireProtocol.download_ext_handler_pkg") as mock_download_ext_handler_pkg:
+            self.ext_handler_instance.download()
+
+        mock_download_ext_handler_pkg.assert_not_called()
+
+        self.assertTrue(os.path.exists(self._get_extension_command_file()), "The extension package was not expanded to the expected location")
+
+    def test_it_should_ignore_existing_extension_package_when_it_is_invalid(self):
+        def download_ext_handler_pkg(_uri, destination):
+            DownloadExtensionTestCase._create_zip_file(destination)
+            return True
+
+        DownloadExtensionTestCase._create_invalid_zip_file(self._get_extension_package_file())
+
+        with patch("azurelinuxagent.common.protocol.wire.WireProtocol.download_ext_handler_pkg", side_effect=download_ext_handler_pkg) as mock_download_ext_handler_pkg:
+            self.ext_handler_instance.download()
+
+        mock_download_ext_handler_pkg.assert_called_once()
+
+        self._assert_download_and_expand_succeeded()
+
+    def test_it_should_use_alternate_uris_when_download_fails(self):
+        self.download_failures = 0
+
+        def download_ext_handler_pkg(_uri, destination):
+            if self.download_failures < 3:
```

Comment:
Can we use NUMBER_OF_DOWNLOAD_RETRIES - 2 instead of 3? It seems like an arbitrary number otherwise.

Reply:
Or is this related to the number of uris you initialize in the test case set up method? In that case, I would also refer to len(self.pkg.uris).

Reply:
It is an arbitrary number :) - Read it as "fail a few times, then succeed".

Reply:
added comment to code